src/content/docs/guides/Beginners/Your First Shader/3_composite_lighting.mdx

```diff
@@ -0,0 +1,134 @@
+vec3 encodedNormal = texture(colortex2, texcoord).rgb;
+vec3 normal = normalize((encodedNormal - 0.5) * 2.0); // we normalize to make sure it is of unit length
+```
+
+Next, let's verify everything is being decoded correctly.
+```glsl
+color.rgb = vec3(lightmap, 0.0);
+```
+![](/src/assets/first_shader_guide/3/lightmap.png)
+
+You can see that where there is skylight, only the green component is set. However, where there is blocklight (by the torch), the colour is yellow. Since red and green make yellow in RGB, we know that blocklight is stored in the red component and sunlight is stored in the green component.
+
+```glsl
+color.rgb = normal;
+```
+![](/src/assets/first_shader_guide/3/normals.png)
+Yep, that looks pretty *normal*. You'll notice that the sky is black here. This is because the `gbuffers_terrain` program does not run for the sky, so no data is stored for these pixels. We will resolve this later on.
+
+## Lighting
+To apply lighting to our colour, we want to multiply it by the colour of the light hitting it. Of course, with multiple light sources, there are multiple types of light hitting it. For our shader, we will have the following types of light
+- Ambient (some constant factor, so you can see in caves)
+- Skylight (blue in the day, uses the lightmap)
+- Blocklight (some warm colour that matches torches, uses the lightmap)
+- Sun/Moonlight (yellow for the sun, white for the moon)
+
+So, to apply our lighting, we would then do something like `color.rgb *= ambient + skylight + blocklight + sunlight`.
+
+First, let's define some colours/brightnesses for each term. I chose the following on a whim.
+```glsl
+const vec3 torchColor = vec3(1.0, 0.5, 0.08);
+const vec3 skyColor = vec3(0.05, 0.15, 0.3);
+const vec3 sunlightColor = vec3(1.0);
+const vec3 ambientColor = vec3(0.1);
+```
+
+I have defined these before `main`, they are `const`s which means the definition can run outside of a the main function.
+
+Let's get these values applied.
+
+```glsl
+vec3 blocklight = lightmap.r * torchColor;
+vec3 skylight = lightmap.g * skyColor;
+vec3 ambient = ambientColor;
+vec3 sunlight = sunlightColor; // we will fix this in a minute
+
+color.rgb *= blocklight + skylight + ambient + sunlight;
+```
+
+Things should look something like this:
+![](/src/assets/first_shader_guide/3/lighting1.png)
+
+## Sunlight
+Now, what about that sunlight? Well, if something is facing directly towards the sun, then we want it to be fully sunlit. On the other hand, if something is facing away from the sun, we want there to be no sunlight. So, we need a function that returns 1.0 if two vectors are facing in the same direction, and 0.0 if they are facing away from each other. Happily, we can use the dot product for this.
+
+Now, the first thing we need to know is where the sun (or moon) is. We can do this with `uniform vec3 shadowLightPosition`. In the daytime this returns the sun's position, in the night, this returns the moon's position.
```

Comment:
What if you split this and did separate sun and moon light and color, would that work?

Reply:
Planned for the later tutorial involving fog and a sky gradient.